consensus/hotstuff/vote_aggregator.go

```diff
@@ -2,27 +2,39 @@ package hotstuff
 
 import (
 	"github.com/onflow/flow-go/consensus/hotstuff/model"
-	"github.com/onflow/flow-go/model/flow"
 )
 
-// VoteAggregator aggregates votes and produces quorum certificates.
+// createVote holds the data for signing a block. The constructor of this function should have the block
+// info already, taking block again as argument is to ensure it's called with the same block.
+// For consensus cluster, the createVote function will create a disposable stateful signer object
+// for producing the signature for a vote. After the the signature has been produced, the signer object will
+// be discarded.
+type createVote func(*model.Block) (*model.Vote, error)
+
+// VoteAggregator verifies, aggregates votes, as well as votes for blocks.
+// When enough votes have been collected, it builds a QC and send it to the EventLoop
+// VoteAggregator also detects protocol violation, including invalid votes, double voting etc, and
+// notifies a HotStuff consumer for slashing.
 type VoteAggregator interface {
 
-	// StorePendingVote is used to store a vote for a block for which we don't
-	// have a block yet.
-	StorePendingVote(vote *model.Vote) (bool, error)
+	// AddVote verifies and aggregates a vote.
+	// The voting block could either be known or unknown.
+	// If the voting block is unknown, the vote won't be processed until AddBlock is called with the block.
+	// This method can be called concurrently, votes will be queued and processed asynchronously.
+	AddVote(vote *model.Vote) error
```

Comment:
Cryptographic audits (and general security) become easier if it's easy to grok when the verification of an individual signature component of the aggregate is done.

Consider `VerifyAndAddVote`?

Reply:
when we receive a vote before its block, we can't verify the vote yet. And we call such vote `PendingVote`.

VoteAggregator and VoteCollector don't verify votes until the block has been received.

`VerifyAndAddVote` would sound like it will verify the vote, but actually VoteAggregator only puts the vote into a queue, never verify. And for vote collector, it verifies only if the block has been received.

But I got your point, we could consider using this name for `VerifyingVoteCollector` cc @huitseeker